File: crates/pray-core/src/registry.rs
use crate::constraint::version_satisfies;
use crate::derived_metadata::RegistryDerivedMetadata;
use crate::manifest::ManifestPackage;
use crate::package_integrity::{artifact_content_digest, require_remote_integrity_fields};
use crate::paths::remove_path_if_exists;
use crate::registry_http::{http_get, http_post, http_put, join_url};
use crate::registry_ssh::{
    resolve_ssh_registry_package_root, submit_confession_ssh, upload_registry_artifact_ssh,
};
use crate::registry_torrent::{fetch_torrent_artifact, fetch_torrent_manifest};
use crate::resolve_context::PackageResolutionContext;
use crate::{PrayError, PrayResult};
use semver::Version;
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone)]
pub struct RegistryPackageResolution {
    pub root: PathBuf,
    pub signer_fingerprint: Option<String>,
    /// Highest non-yanked version published in registry metadata, regardless of Prayfile constraint.
    pub registry_latest_version: Option<String>,
}

pub fn lockfile_signer_fingerprint(version: &RegistryPackageVersion) -> Option<String> {
    version
        .signer_fingerprint
        .as_deref()
        .filter(|value| crate::ssh_identity::looks_like_ssh_fingerprint(value))
        .map(crate::ssh_identity::normalize_identity)
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub struct RegistryIndex {
    pub spec: String,
    #[serde(default)]
    pub packages: Vec<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub struct RegistryPackageMetadata {
    pub name: String,
    #[serde(default)]
    pub versions: Vec<RegistryPackageVersion>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub struct RegistryPackageVersion {
    pub version: String,
    pub artifact: String,
    #[serde(default)]
    pub artifact_hash: Option<String>,
    #[serde(default)]
    pub tree_hash: Option<String>,
    #[serde(default)]
    pub yanked: bool,
    #[serde(default)]
    pub targets: Vec<String>,
    #[serde(default)]
    pub exports: Vec<String>,
    #[serde(default)]
    pub signer: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub signer_fingerprint: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub signer_public_key: Option<String>,
    #[serde(default)]
    pub published_at: Option<String>,
    #[serde(default)]
    pub signature: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub derived_metadata: Option<RegistryDerivedMetadata>,
}

impl RegistryPackageVersion {
    pub fn same_identity(&self, other: &Self) -> bool {
        self.version == other.version
            && self.artifact == other.artifact
            && self.artifact_hash == other.artifact_hash
            && self.tree_hash == other.tree_hash
            && self.yanked == other.yanked
            && self.targets == other.targets
            && self.exports == other.exports
            && self.signer == other.signer
            && self.signer_fingerprint == other.signer_fingerprint
            && self.signer_public_key == other.signer_public_key
            && self.published_at == other.published_at
            && self.signature == other.signature
    }

    pub fn merge_annotations_from(&mut self, other: &Self) {
        if self.derived_metadata.is_none() {
            self.derived_metadata = other.derived_metadata.clone();
        }
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub struct ConfessionSubmission {
    pub package: String,
    pub version: String,
    pub status: String,
    #[serde(default)]
    pub note: Option<String>,
    #[serde(default)]
    pub lockfile: Option<String>,
    #[serde(default)]
    pub distribution_point: Option<String>,
    #[serde(default)]
    pub signer: Option<String>,
    #[serde(default)]
    pub timestamp: Option<String>,
    #[serde(default)]
    pub signature: Option<String>,
}

pub fn resolve_registry_package_root(
    project_root: &Path,
    source_url: &str,
    declaration: &ManifestPackage,
    context: &PackageResolutionContext,
) -> PrayResult<RegistryPackageResolution> {
    if crate::ssh_client::is_pray_ssh_url(source_url) {
        return resolve_ssh_registry_package_root(project_root, source_url, declaration, context);
    }

    let metadata = fetch_registry_package_metadata(source_url, &declaration.name)?;
    let registry_latest_version = registry_latest_version_label(&metadata);
    let selected = select_package_version(
        &metadata,
        &declaration.constraint,
        context.preferred_version.as_deref(),
    )?;
    let signer_fingerprint = lockfile_signer_fingerprint(&selected);
    require_remote_integrity_fields(&declaration.name, &selected.version, &selected)?;
    if let Some(vendored_root) = crate::registry_cache::try_vendored_package_root(
        project_root,
        &declaration.name,
        &selected,
    )? {
        return Ok(RegistryPackageResolution {
            root: vendored_root,
            signer_fingerprint,
            registry_latest_version,
        });
    }
    let cache_directory = crate::registry_cache::registry_cache_directory(
        project_root,
        source_url,
        &declaration.name,
        &selected.version,
    );

    if let Some(mut cached) = crate::registry_cache::try_reuse_cached_registry_package(
        &cache_directory,
        &selected,
        signer_fingerprint.clone(),
    )? {
        cached.registry_latest_version = registry_latest_version.clone();
        return Ok(cached);
    }
    if context.offline {
        return Err(crate::registry_cache::offline_package_error(
            &declaration.name,
            &selected.version,
        ));
    }

    if cache_directory.exists() {
        remove_path_if_exists(&cache_directory)?;
    }
    fs::create_dir_all(&cache_directory)?;

    let artifact_url = join_url(source_url, &selected.artifact);
    let torrent_manifest = fetch_torrent_manifest(source_url, &selected.artifact)?;
    let artifact_bytes = if let Some(manifest) = torrent_manifest {
        fetch_torrent_artifact(source_url, &selected.artifact, &manifest)?
    } else {
        http_get(&artifact_url)?
    };
    crate::registry_cache::validate_and_unpack_registry_package(
        &cache_directory,
        declaration,
        &selected,
        &artifact_bytes,
    )?;

    Ok(RegistryPackageResolution {
        root: cache_directory,
        signer_fingerprint,
        registry_latest_version,
    })
}

pub fn resolve_local_registry_package_root(
    project_root: &Path,
    source_key: &str,
    source_root: &Path,
    declaration: &ManifestPackage,
    context: &PackageResolutionContext,
) -> PrayResult<RegistryPackageResolution> {
    let metadata_path = source_root.join(format!("v1/packages/{}.json", declaration.name));
    let metadata_text = fs::read_to_string(&metadata_path).map_err(|error| {
        if error.kind() == std::io::ErrorKind::NotFound {
            PrayError::Resolution(format!(
                "package {} not found in distribution {:?}. \
                 Missing {}. Check the package name, version constraint `{}`, and that the source publishes registry metadata.",
                declaration.name,
                source_root,
                metadata_path.display(),
                declaration.constraint
            ))
        } else {
            PrayError::Resolution(format!(
                "failed to read package metadata {}: {error}",
                metadata_path.display()
            ))
        }
    })?;
    let metadata: RegistryPackageMetadata =
        serde_json::from_str(&metadata_text).map_err(|error| PrayError::Parse {
            kind: "registry metadata",
            message: error.to_string(),
        })?;
    let registry_latest_version = registry_latest_version_label(&metadata);
    let selected = select_package_version(
        &metadata,
        &declaration.constraint,
        context.preferred_version.as_deref(),
    )?;
    let signer_fingerprint = lockfile_signer_fingerprint(&selected);
    require_remote_integrity_fields(&declaration.name, &selected.version, &selected)?;
    if let Some(vendored_root) = crate::registry_cache::try_vendored_package_root(
        project_root,
        &declaration.name,
        &selected,
    )? {
        return Ok(RegistryPackageResolution {
            root: vendored_root,
            signer_fingerprint,
            registry_latest_version,
        });
    }
    let cache_identifier = format!(
        "{}:{}:{}:{}",
        source_key,
        declaration.name,
        selected.version,
        selected
            .artifact_hash
            .as_deref()
            .unwrap_or("no-artifact-hash")
    );
    let cache_directory = crate::registry_cache::registry_cache_directory(
        project_root,
        &cache_identifier,
        &declaration.name,
        &selected.version,
    );

    if let Some(mut cached) = crate::registry_cache::try_reuse_cached_registry_package(
        &cache_directory,
        &selected,
        signer_fingerprint.clone(),
    )? {
        cached.registry_latest_version = registry_latest_version.clone();
        return Ok(cached);
    }
    if context.offline {
        return Err(crate::registry_cache::offline_package_error(
            &declaration.name,
            &selected.version,
        ));
    }

    if cache_directory.exists() {
        remove_path_if_exists(&cache_directory)?;
    }
    fs::create_dir_all(&cache_directory)?;

    let artifact_bytes =
        crate::registry_cache::read_local_registry_artifact_bytes(source_root, &selected.artifact)?;
    crate::registry_cache::validate_and_unpack_registry_package(
        &cache_directory,
        declaration,
        &selected,
        &artifact_bytes,
    )?;

    Ok(RegistryPackageResolution {
        root: cache_directory,
        signer_fingerprint,
        registry_latest_version,
    })
}

pub fn registry_package_signing_identity(version: &RegistryPackageVersion) -> Option<String> {
    crate::ssh_identity::package_signing_identity(
        version.signer.as_deref(),
        version.signer_fingerprint.as_deref(),
    )
}

pub fn registry_artifact_signature(artifact_bytes: &[u8], tree_hash: &str, signer: &str) -> String {
    artifact_content_digest(artifact_bytes, tree_hash, signer)
}

pub fn submit_confession(source_url: &str, confession: &ConfessionSubmission) -> PrayResult<()> {
    if crate::ssh_client::is_pray_ssh_url(source_url) {
        return submit_confession_ssh(source_url, confession);
    }
    let endpoint = join_url(source_url, "v1/confessions");
    let payload =
        serde_json::to_vec(confession).map_err(|error| PrayError::Manifest(error.to_string()))?;
    let response = http_post(&endpoint, "application/json", &payload)?;
    if response.status / 100 != 2 {
        return Err(PrayError::Resolution(format!(
            "confession submission failed with HTTP {}",
            response.status
        )));
    }
    Ok(())
}

pub fn upload_registry_artifact(
    source_url: &str,
    artifact_path: &str,
    bytes: &[u8],
) -> PrayResult<()> {
    if crate::ssh_client::is_pray_ssh_url(source_url) {
        return upload_registry_artifact_ssh(source_url, artifact_path, bytes);
    }
    let endpoint = join_url(source_url, artifact_path);
    let response = http_put(&endpoint, "application/octet-stream", bytes)?;
    if response.status / 100 != 2 {
        return Err(PrayError::Resolution(format!(
            "artifact upload failed with HTTP {}",
            response.status
        )));
    }
    Ok(())
}

fn fetch_registry_package_metadata(
    source_url: &str,
    package_name: &str,
) -> PrayResult<RegistryPackageMetadata> {
    let url = join_url(source_url, &format!("v1/packages/{}.json", package_name));
    let response = http_get(&url)?;
    serde_json::from_slice(&response).map_err(|error| PrayError::Parse {
        kind: "registry metadata",
        message: error.to_string(),
    })
}

pub fn highest_registry_version(
    metadata: &RegistryPackageMetadata,
) -> PrayResult<Option<RegistryPackageVersion>> {
    let mut selected: Option<RegistryPackageVersion> = None;
    for version in &metadata.versions {
        if version.yanked {
            continue;
        }
        match &selected {
            Some(existing) if compare_versions(&version.version, &existing.version)? <= 0 => {}
            _ => selected = Some(version.clone()),
        }
    }
    Ok(selected)
}

pub fn registry_latest_version_label(metadata: &RegistryPackageMetadata) -> Option<String> {
    highest_registry_version(metadata)
        .ok()
        .flatten()
        .map(|version| version.version)
}

pub fn version_is_greater_than(left: &str, right: &str) -> PrayResult<bool> {
    Ok(compare_versions(left, right)? > 0)
}

pub(crate) fn select_package_version(
    metadata: &RegistryPackageMetadata,
    constraint: &str,
    preferred_version: Option<&str>,
) -> PrayResult<RegistryPackageVersion> {
    if let Some(preferred_version) = preferred_version {
        if let Some(version) = metadata
            .versions
            .iter()
            .find(|version| version.version == preferred_version && !version.yanked)
        {
            if version_satisfies(&version.version, constraint)? {
                return Ok(version.clone());
            }
            // Prayfile constraint changed; fall through to the highest satisfying version.
        }
    }
    let mut selected: Option<RegistryPackageVersion> = None;
    for version in &metadata.versions {
        if version.yanked {
            continue;
        }
        if !version_satisfies(&version.version, constraint)? {
            continue;
        }
        match &selected {
            Some(existing) if compare_versions(&version.version, &existing.version)? <= 0 => {}
            _ => selected = Some(version.clone()),
        }
    }
    selected.ok_or_else(|| {
        PrayError::Resolution(format!(
            "no registry version for {} satisfies {}",
            metadata.name, constraint
        ))
    })
}

#[doc(hidden)]
pub fn select_package_version_for_test(
    metadata: &RegistryPackageMetadata,
    constraint: &str,
    preferred_version: Option<&str>,
) -> PrayResult<RegistryPackageVersion> {
    select_package_version(metadata, constraint, preferred_version)
}

fn compare_versions(left: &str, right: &str) -> PrayResult<i32> {
    let left = Version::parse(left).map_err(|error| PrayError::Resolution(error.to_string()))?;
    let right = Version::parse(right).map_err(|error| PrayError::Resolution(error.to_string()))?;
    Ok(match left.cmp(&right) {
        std::cmp::Ordering::Less => -1,
        std::cmp::Ordering::Equal => 0,
        std::cmp::Ordering::Greater => 1,
    })
}

pub fn fetch_optional_distribution_bytes(
    source_url: &str,
    relative_path: &str,
) -> PrayResult<Option<Vec<u8>>> {
    if !source_url.starts_with("http://") && !source_url.starts_with("https://") {
        return Err(PrayError::Unsupported(format!(
            "distribution fetch requires http or https source, got {source_url}"
        )));
    }
    let url = join_url(source_url, relative_path);
    match http_get(&url) {
        Ok(bytes) => Ok(Some(bytes)),
        Err(PrayError::Resolution(message)) if message.contains("404") => Ok(None),
        Err(error) => Err(error),
    }
}

File: rubygems/pray-cli/lib/pray/registry.rb
# frozen_string_literal: true

require "json"
require "net/http"
require "uri"
require "fileutils"
require "pathname"
require_relative "path_safety"

module Pray
  RegistryPackageVersion = Struct.new(
    :version, :artifact, :artifact_hash, :tree_hash, :yanked, :targets, :exports,
    :signer, :signer_fingerprint, :published_at, :signature
  ) do
    def initialize(
      artifact_hash: nil, tree_hash: nil, yanked: false, targets: [], exports: [],
      signer: nil, signer_fingerprint: nil, published_at: nil, signature: nil, **kwargs
    )
      super(**kwargs, artifact_hash: artifact_hash, tree_hash: tree_hash, yanked: yanked,
                      targets: targets, exports: exports, signer: signer,
                      signer_fingerprint: signer_fingerprint, published_at: published_at, signature: signature)
    end
  end

  RegistryPackageMetadata = Struct.new(:name, :versions) do
    def initialize(name: nil, versions: [])
      super
    end
  end

  RegistryPackageResolution = Struct.new(
    :root, :signer_fingerprint, :registry_latest_version
  )

  module Registry
    module_function

    def resolve_registry_package_root(project_root, source_url, declaration, preferred_version: nil, offline: false)
      metadata = fetch_package_metadata(source_url, declaration.name)
      registry_latest_version = registry_latest_version_label(metadata)
      selected = select_package_version(metadata, declaration.constraint, preferred_version)
      cache_directory = registry_cache_directory(
        project_root,
        source_url,
        declaration.name,
        selected.version,
        selected.artifact_hash
      )

      if cache_ready?(cache_directory, selected)
        return RegistryPackageResolution.new(
          root: cache_directory,
          signer_fingerprint: selected.signer_fingerprint,
          registry_latest_version: registry_latest_version
        )
      end

      raise Error.resolution(offline_package_error(declaration.name, selected.version)) if offline

      FileUtils.rm_rf(cache_directory) if File.exist?(cache_directory)
      FileUtils.mkdir_p(cache_directory)

      artifact_bytes = read_artifact_bytes(source_url, selected.artifact)
      validate_and_unpack(cache_directory, declaration, selected, artifact_bytes, source_url: source_url)

      RegistryPackageResolution.new(
        root: cache_directory,
        signer_fingerprint: selected.signer_fingerprint,
        registry_latest_version: registry_latest_version
      )
    end

    def resolve_local_registry_package_root(project_root, source_key, source_root, declaration, preferred_version: nil, offline: false)
      metadata_path = registry_metadata_path(source_root, declaration.name)
      unless File.exist?(metadata_path)
        raise Error.resolution(
          "package #{declaration.name} not found in distribution #{source_root.inspect}. " \
          "Missing #{metadata_path}. Check the package name, version constraint `#{declaration.constraint}`, " \
          "and that the source publishes registry metadata."
        )
      end

      metadata = parse_metadata(File.read(metadata_path, encoding: "UTF-8"))
      registry_latest_version = registry_latest_version_label(metadata)
      selected = select_package_version(metadata, declaration.constraint, preferred_version)
      cache_directory = registry_cache_directory(
        project_root,
        source_key,
        declaration.name,
        selected.version,
        selected.artifact_hash
      )

      if cache_ready?(cache_directory, selected)
        return RegistryPackageResolution.new(
          root: cache_directory,
          signer_fingerprint: selected.signer_fingerprint,
          registry_latest_version: registry_latest_version
        )
      end

      raise Error.resolution(offline_package_error(declaration.name, selected.version)) if offline

      FileUtils.rm_rf(cache_directory) if File.exist?(cache_directory)
      FileUtils.mkdir_p(cache_directory)

      artifact_bytes = read_local_artifact_bytes(source_root, selected.artifact)
      validate_and_unpack(cache_directory, declaration, selected, artifact_bytes, source_url: source_root)

      RegistryPackageResolution.new(
        root: cache_directory,
        signer_fingerprint: selected.signer_fingerprint,
        registry_latest_version: registry_latest_version
      )
    end

    def fetch_package_metadata(source_url, package_name)
      if local_source?(source_url)
        metadata_path = registry_metadata_path(source_url, package_name)
        return parse_metadata(File.read(metadata_path, encoding: "UTF-8"))
      end

      PathSafety.reject_unsafe_package_name!(package_name)
      response = http_get(join_url(source_url, "v1/packages/#{package_name}.json"))
      parse_metadata(response)
    end

    def parse_metadata(text)
      data = JSON.parse(text)
      RegistryPackageMetadata.new(
        name: data["name"],
        versions: Array(data["versions"]).map { |entry| version_from_hash(entry) }
      )
    end

    def version_from_hash(entry)
      RegistryPackageVersion.new(
        version: entry["version"],
        artifact: entry["artifact"],
        artifact_hash: entry["artifact_hash"],
        tree_hash: entry["tree_hash"],
        yanked: entry["yanked"] || false,
        targets: entry["targets"] || [],
        exports: entry["exports"] || [],
        signer: entry["signer"],
        signer_fingerprint: entry["signer_fingerprint"],
        published_at: entry["published_at"],
        signature: entry["signature"]
      )
    end

    def select_package_version(metadata, constraint, preferred_version)
      if preferred_version
        version = metadata.versions.find { |entry| entry.version == preferred_version && !entry.yanked }
        return version if version && Constraint.version_satisfies(version.version, constraint)
      end

      selected = nil
      metadata.versions.each do |version|
        next if version.yanked
        next unless Constraint.version_satisfies(version.version, constraint)

        selected = version if selected.nil? || compare_versions(version.version, selected.version).positive?
      end

      raise Error.resolution("no registry version for #{metadata.name} satisfies #{constraint}") unless selected

      selected
    end

    def registry_latest_version_label(metadata)
      highest = metadata.versions.reject(&:yanked).max_by { |entry| Gem::Version.new(entry.version) }
      highest&.version
    end

    def compare_versions(left, right)
      Gem::Version.new(left) <=> Gem::Version.new(right)
    end

    def validate_and_unpack(cache_directory, declaration, selected, artifact_bytes, source_url: nil)
      if selected.artifact_hash
        artifact_hash = Hashing.sha256_prefixed(artifact_bytes)
        if artifact_hash != selected.artifact_hash
          raise Error.integrity(
            "package artifact hash mismatch for #{declaration.name} #{selected.version}"
          )
        end
      end

      verify_registry_signature!(declaration, selected, artifact_bytes)
      Trust.verify_publisher_fingerprint!(source_url, selected) if source_url

      Archive.unpack_praypkg(artifact_bytes, cache_directory)
      spec_path = Resolve.find_prayspec_file(cache_directory)
      spec = Pray.parse_package_spec(File.read(spec_path)).canonicalized
      if spec.name != declaration.name
        raise Error.resolution(
          "package path #{cache_directory.inspect} declares #{spec.name.inspect}, expected #{declaration.name.inspect}"
        )
      end
      if spec.version != selected.version
        raise Error.resolution(
          "package #{declaration.name} version #{spec.version} does not match registry version #{selected.version}"
        )
      end
      if selected.tree_hash
        actual_tree_hash = spec.tree_hash_for_root(cache_directory)
        if actual_tree_hash != selected.tree_hash
          raise Error.integrity(
            "package tree hash mismatch for #{declaration.name} #{selected.version}"
          )
        end
      end
    end

    def read_local_artifact_bytes(source_root, artifact)
      if artifact.start_with?("http://", "https://")
        return http_get(artifact)
      end
      if artifact.start_with?("file://")
        path = PathSafety.join_under_root(source_root, artifact.delete_prefix("file://"))
        raise Error.resolution("package artifact path escapes distribution root") unless path
        return File.binread(path)
      end

      path = PathSafety.join_under_root(source_root, artifact)
      raise Error.resolution("package artifact path escapes distribution root") unless path
      raise Error.resolution("package artifact missing at #{path}") unless File.exist?(path)

      File.binread(path)
    end

    def read_artifact_bytes(source_url, artifact)
      return read_local_artifact_bytes(source_url, artifact) if local_source?(source_url)

      http_get(join_url(source_url, artifact))
    end

    def cache_ready?(cache_directory, selected)
      return false unless File.directory?(cache_directory)

      spec_path = Resolve.find_prayspec_file(cache_directory)
      spec = Pray.parse_package_spec(File.read(spec_path)).canonicalized
      spec.version == selected.version
    rescue Errno::ENOENT, Errno::ENOTDIR, SystemCallError
      false
    end

    def registry_metadata_path(source_root, package_name)
      PathSafety.reject_unsafe_package_name!(package_name)
      packages_root = PathSafety.join_under_root(source_root, "v1", "packages")
      raise Error.resolution("invalid package name: #{package_name.inspect}") unless packages_root

      metadata_path = Pathname.new(packages_root).join("#{package_name}.json").cleanpath.to_s
      unless PathSafety.path_under_root?(packages_root, metadata_path)
        raise Error.resolution("invalid package name: #{package_name.inspect}")
      end

      metadata_path
    end

    def verify_registry_signature!(declaration, selected, artifact_bytes)
      return unless selected.signature

      unless selected.signer && selected.tree_hash
        raise Error.integrity(
          "package signature metadata incomplete for #{declaration.name} #{selected.version}"
        )
      end

      expected = registry_artifact_signature(artifact_bytes, selected.tree_hash, selected.signer)
      return if expected == selected.signature

      raise Error.integrity(
        "package signature mismatch for #{declaration.name} #{selected.version}"
      )
    end

    def registry_cache_directory(project_root, source_key, package_name, version, artifact_hash)
      identifier = [
        source_key,
        package_name,
        version,
        artifact_hash || "no-artifact-hash"
      ].join(":")
      digest = Hashing.sha256_hex(identifier)[0, 16]
      File.join(project_root, ".pray", "cache", "registry", package_name.tr("/", "-"), version, digest)
    end

    def offline_package_error(package_name, version)
      "package #{package_name} #{version} is not cached and offline mode is enabled"
    end

    def local_source?(source_url)
      !source_url.start_with?("http://", "https://", "pray+ssh://", "ssh+pray://")
    end

    def join_url(base, path)
      URI.join(base.end_with?("/") ? base : "#{base}/", path.delete_prefix("/")).to_s
    end

    def http_get(url)
      uri = URI(url)
      response = http_request(uri) { |http| http.get(uri.request_uri) }
      unless response.is_a?(Net::HTTPSuccess)
        raise Error.resolution("HTTP request failed for #{url}: #{response.code}")
      end

      response.body
    end

    def http_put(url, content_type, body)
      uri = URI(url)
      request = Net::HTTP::Put.new(uri)
      request["Content-Type"] = content_type
      request.body = body
      response = http_request(uri) { |http| http.request(request) }
      unless response.is_a?(Net::HTTPSuccess)
        raise Error.resolution("HTTP upload failed for #{url}: #{response.code}")
      end

      response.body
    end

    def http_request(uri)
      Net::HTTP.start(
        uri.hostname,
        uri.port,
        use_ssl: uri.scheme == "https",
        open_timeout: 10,
        read_timeout: 30
      ) do |http|
        yield http
      end
    end

    def registry_artifact_signature(artifact_bytes, tree_hash, signer)
      payload = artifact_bytes + "\0".b + tree_hash.b + "\0".b + signer.b
      Hashing.sha256_prefixed(payload)
    end
  end
end
